cpp/src/arrow/json/from_string.h

```diff
@@ -0,0 +1,88 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+// Implement a simple JSON representation format for arrays
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <string_view>
+
+#include "arrow/status.h"
+#include "arrow/type_fwd.h"
+#include "arrow/util/visibility.h"
+
+namespace arrow {
+
+class Array;
+class DataType;
+
+namespace json {
+
+/// \defgroup array-from-json-string Helpers for constructing Arrays from JSON
+/// text
+///
+/// These helpers are intended to be used in examples, tests, or for quick
+/// prototyping and are not intended to be used where performance matters.
+///
+/// @{
+
+/// \brief Create an Array from a JSON string
```

Comment:
It'd be nice if these included at least one usage example

```suggestion
/// \brief Create an Array from a JSON string
///
/// \code {.cpp}
/// std::shared_ptr<Array> array = ArrayFromJSONString(
///   int64(), "[2, 3, 5, 7, 11]"
/// ).ValueOrDie();
/// \endcode
```

Reply:
Yeah this is a good point. I'll document these better.

Reply:
I added a complete listing in e4e63af21afec18d8fafab6c537e8c2a1778c732.